policy: Add framework for async event completion notifications.

jrajahalme opened the pull request:
Add a framework for asynchronous proxy update completion notifications and use it for Envoy.

rlenglet commented on pkg/policy/policy.go:
```diff
@@ -26,6 +27,18 @@ import (
+	Completed(bool)
+}
+
+// CompletionContainer is an interface for pooling asynchronous success/failure
```
pooling -> polling (?)

jrajahalme replied:
Pooling is the right spelling here :-)

Multiple completions can be added to the same container, so that the caller can efficiently wait for multiple compoletions at one time.

rlenglet commented on pkg/envoy/streamctrl.go:
```diff
@@ -38,6 +46,58 @@ type StreamControl struct {
+func (ctrl *StreamControl) handleCompletions(version uint64, success bool) {
+	now := time.Now()
+
+	for idx := 0; idx < len(ctrl.completions); {
```
```
for _, completion := range ctrl.completions {
    ...
}
```

jrajahalme replied:
Does that support removal of entries in the slice while iterating?

rlenglet commented on pkg/policy/policy.go:
```diff
@@ -26,6 +27,18 @@ import (
+// CompletionContainer is an interface for pooling asynchronous success/failure
+// notifications.
+type CompletionContainer interface {
+	AddCompletion() (Completion, time.Duration)
```
Instead of explicitly using a timeout, why not pass a `context.Context` to every calls that currently takes a `CompletionContainer`? Using a context will allow your streamctrl logic to support both explicit cancelations from callers and timeouts.
Instead of directly dealing with timeouts, you'd only need to check whether the context is still valid with `ctx.Err() == nil` to determine whether to continue waiting for an ACK.

rlenglet replied:
Also, all functions that currently take a `CompletionContainer` as a parameter only either pass it down the call stack, or call `AddCompletion` only once. I'd directly pass a `Completion` parameter instead of a `CompletionContainer`.

It may also be worth trying to implement the Future pattern, because that's really what you try to do here. A future may simply be modeled with a `chan bool` here. The callee would write at most one value into the channel and close it, once the version is ACKed/NACKed by the proxy.
The functions' interfaces would then look like:
```
func (s *LDSServer) updateListener(ctx context.Context, ..., out chan<- bool) {
    ...
}
```

If you wanted to define a fancy type to wrap the channel details, you could pass a `Completion` to do the job, and implement it so:
```
type Completion func(success bool)

func NewCompletion(ctx Context, out chan<- bool) Completion {
    return func(success bool) {
        select {
        case <-ctx.Done():
            // Ignore.
        case out <- success:
        }
        close(out)
    }
}

func (s *LDSServer) updateListener(ctx context.Context, ..., completion Completion) {
    ...
}
```

rlenglet replied:
Now I think you'll always want to combine a completion with a context, so it may be better to define the interface so:
```
type Completion interface {
    Context() context.Context
    Completed(success bool)
}
```

And just pass only that around, e.g.:
```
func (e *Envoy) UpdateListener(name string, l7rules policy.L7DataMap, completion policy.Completion) {
    ...
}
```

jrajahalme replied:
Part of my design was to both:
- Allow passing CompletionContainer as nil to support callers who may not be interested in the completion signal
- Allow implementations that are always synchronous to ignore the CompletionContainer

Furthermore, my implementation allows:
- the implementation of a Completion interface package a single WaitGroup
- and share that same instance for multiple Completions in the same CompletionsContainer

This allows for a potentially large number of completions to be waited for with a single WaitGroup.Wait().

To do the same with a single channel without blocking writers I'd have to have a buffered channel for some max number of completions?

jrajahalme replied:
I'll have to look at the timeout.. Maybe it would be better to have an overall daedline for the whole policy update operartion rather than individual timeouts on each completion. If so the Context() could be a member in the CompletionContainer interface. I guess context.Context must be passed by reference for a cancellation to pass through?

rlenglet commented on pkg/policy/policy.go:
```diff
@@ -26,6 +27,18 @@ import (
 
+// Completion is an interface for an individual asynchronous success/failure
+// notification.
+type Completion interface {
```
This could simply be defined as a func:
```
type Completion func(success bool)
```

jrajahalme replied:
An interface carries an object reference with it, while a function pointer does not. Multiple different endpoints may request completion callbacks in parallel from the same stream (in case of LDS).

rlenglet reviewed (changes requested):
I have not reviewed all yet. I have general comments about the design of Completion / CompletionContainer.

jrajahalme reviewed (commented):
Thanks for the review! See my responses :-)

nebril commented on pkg/envoy/streamctrl.go:
```diff
@@ -38,6 +46,58 @@ type StreamControl struct {
+		log.Debug("Envoy: ", ctrl.name, " ", comp.msg, " ", res, ", time left: ", comp.expiration.Sub(now))
+
+		// Remove the entry
+		ctrl.completions[idx] = ctrl.completions[len(ctrl.completions)-1]
```
This is a weird construction IMO, you are effectively shrinking `ctrl.completions` slice until you processed all entries.

If `ctrl.completions` can be changed after going through this loop I would suggest to rewrite this whole loop as follows, to avoid somewhat convoluted slice filtering:
```
	tmpCompletions := make([]versionCompletion, 0, len(ctrl.completions))
	for _, comp := range ctrl.completions {

		result := success

		if version >= comp.version {
			// Ack or Nack for this version OR later received
		} else if now.After(comp.expiration) {
			// Timed out, return failure.
			result = false
		} else {
			// retain the entry
			append(tmpCompletions, comp)
		}

		res := "NACK"
		if result {
			res = "ACK"
		}
		log.Debug("Envoy: ", ctrl.name, " ", comp.msg, " ", res, ", time left: ", comp.expiration.Sub(now))

		comp.completion.Completed(result)
	}
	ctrl.completions = tmpCompletions
```
Hope this makes sense.

jrajahalme replied:
Makes sense, even though I don't like the fact that range loop copies the object into a temporary (here `comp`). However, in most cases we do this after a successful Envoy transaction, so it makes no difference in practice.

I also did not include the preallocation of the temporary array, as with that the array never shrinks. While we are not dealing with large arrays, IMO it is a bad habit. Again, in most cases I expect we don't need to retain entries, so this does not make a difference either.

I also had to keep the `continue` to not call completion when an entry is retained.

I do like that this is more readable now!

nebril reviewed (changes requested):
One loop rewrite suggestion inline, please dismiss review if you don't agree with it.

tgraf commented on pkg/envoy/envoy_test.go:
```diff
@@ -38,7 +38,7 @@ func (s *EnvoySuite) TestEnvoy(c *C) {
 	stateLogDir := c.MkDir()
 
 	// launch debug variant of the Envoy proxy
-	Envoy := StartEnvoy(9901, stateLogDir, stateLogDir, 42)
```
Typo in your email in the comment message

jrajahalme replied:
Fixed.

tgraf commented on pkg/endpoint/policy.go:
```diff
@@ -670,10 +682,13 @@ func (e *Endpoint) regenerate(owner Owner, reason string) (retErr error) {
 	e.BuildMutex.Lock()
 	defer e.BuildMutex.Unlock()
+	e.ProxyCompletions = &Completions{}
+	defer func() { e.ProxyCompletions = nil }()
```
This is very fragile as it depends on the above defer to be called afterwards. Can you do one defer function which sets ProxyCompletions to nil and then unlocks BuildMutex so it's clear that there is a dependency?

jrajahalme replied:
Good point, with one defer the order is more explicit and it becomes a bit more efficient as well.

jrajahalme replied:
Good point, with one defer the order is more explicit and it becomes a bit more efficient as well.

tgraf reviewed (changes requested):
Everybody please remember that this is an expedited bug which is currently blocking several other bugs. I don't mind long term design discussions but was there anything in the original implementation that did *not* work and requires a re-design or can we merge the original approach and then refine later?

jrajahalme committed:
envoy_test: Use different admin port.

Unit test fails if cilium is installed and running Envoy at the same
time due to the port being already in use by the installed and running
cilium-envoy. Use a different admin port to be able to run the unit
test at the same time with the installed cilium-envoy.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
envoy/Makefile: List proto sources explicitly.

Wildcard expands to all existing files in the make target, rather than
just the files we intend. Also, it is better to explicitly list the
files as sometimes old files remain behind when changing from a git
branch to another.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
policy: Add framework for async event completion notifications.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

manalibhutiyani commented on daemon/daemon.go:
```diff
@@ -127,7 +127,7 @@ func (d *Daemon) UpdateProxyRedirect(e *endpoint.Endpoint, l4 *policy.L4Filter)
 	}
 
-	r, err := d.l7Proxy.CreateOrUpdateRedirect(l4, e.ProxyID(l4), e, d)
+	r, err := d.l7Proxy.CreateOrUpdateRedirect(l4, e.ProxyID(l4), e, d, e)
```
Why do we need to pass `e` twice, as well as `e.ProxyID(l4)` separately?

jrajahalme replied:
'e', endpoint is being passed in different roles (literally different interfaces), the called function does not (need to) know that they are the same 'e':

```
func (p *Proxy) CreateOrUpdateRedirect(l4 *policy.L4Filter, id string, source ProxySource,
	notifier accesslog.LogRecordNotifier, completions policy.CompletionContainer) (Redirect, error)
```

rlenglet reviewed (approved):
OK with addressing the interface design in a subsequent PR.
